sdk/core/azure-core-amqp/src/impl/rust_amqp/rust_amqp/rust_wrapper/src/call_context.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corp. All Rights Reserved
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+// cspell: words reqwest repr tokio
+
+use crate::{runtime_context::RuntimeContext, rust_error::RustError};
+
+pub struct RustCallContext {
+    runtime_context: *mut RuntimeContext,
+    error: Option<RustError>,
+}
+
+impl RustCallContext {
+    pub fn new(runtime_context: *mut RuntimeContext) -> Self {
+        Self {
+            runtime_context,
+            error: None,
+        }
+    }
+
+    pub fn runtime_context(&self) -> &RuntimeContext {
+        unsafe { &*self.runtime_context }
+    }
+
+    pub fn set_error(&mut self, error: Box<dyn std::error::Error + Send + Sync>) {
+        self.error = Some(RustError::new(error));
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn call_context_new(runtime_context: *mut RuntimeContext) -> *mut RustCallContext {
+    Box::into_raw(Box::new(RustCallContext::new(runtime_context)))
+}
+
+#[no_mangle]
+pub extern "C" fn call_context_delete(ctx: *mut RustCallContext) {
+    unsafe {
+        drop(Box::from_raw(ctx));
+    }
+}
```

Comment:
Why not just take it as a value and it'll be dropped automatically, rather than boxing it just to drop it? Seems excessive:
```suggestion
pub extern "C" fn call_context_delete(ctx: RustCallContext) {}
```

This is the typical C++<->Rust interop pattern for objects.

Reply:
I can certainly try that - the cbindgen examples I saw all used Box::into_ptr to return values.

Reply:
Rust gets upset because RustCallContext is not FFI safe. And I can't make it FFI safe - basically I want to allocate a pointer to the object and return that pointer to the caller in the xxx_create call and I want to delete the pointer in the xxx_destroy call.

That lets me put native rust types in the structure without having to worry about the structure layout matching anything in C++. And these types are all opaque types to C++ anyway. C++ considers these pointers to declared but not defined types.